### PR TITLE
VersionVector CRDT を実装

### PR DESCRIPTION
**Pull request by j5ik2o**
## 概要

- core/ddata に `VersionVector` と `VersionVectorOrdering` を追加
- node ごとの causal ordering / merge / removed-node pruning を実装
- `VersionVectorKey` と property tests を追加し、cluster gap analysis の Phase 2 進捗を更新

## 検証

- `cargo test -p fraktor-cluster-core-kernel-rs version_vector --lib`
- `./scripts/ci-check.sh ai fmt dylint clippy unit-test`
- `./scripts/ci-check.sh ai all`

---

> [!NOTE]
> **Low Risk**
> New library-only ddata types with no runtime or network wiring; behavior is covered by unit and property tests.
> 
> **Overview**
> Adds a **`VersionVector`** CRDT in `cluster-core-kernel` `ddata` for per-node causal history (Pekko distributed-data parity), with **`VersionVectorOrdering`** for same / before / after / concurrent comparisons.
> 
> The type supports **`increment`**, **`compare`**, **`ReplicatedData::merge`** (per-node max), and **`RemovedNodePruning`** (collapse removed nodes into a survivor with a bumped clock, plus cleanup). **`VersionVectorKey`** is wired into the typed key aliases and public `ddata` exports.
> 
> **Tests** cover ordering, merge, pruning edge cases, and proptest merge laws (commutative, associative, idempotent). **`docs/gap-analysis/cluster-gap-analysis.md`** marks `VersionVector` implemented, bumps overall parity to **105/151 (70%)**, and updates Distributed Data / CRDT category counts and Phase 2 notes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2c9d8d8150cc1ff46e087499da84819e2264079d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Release Notes

* **新機能**
  * VersionVector（分散因果時計）の実装が完了。ノード間の履歴順序判定とデータマージ機能が利用可能になりました。

* **ドキュメント**
  * 分散データ層の実装状況を更新。カバレッジが 69% から 70% へ向上しました。

* **テスト**
  * VersionVector の動作検証テストが追加されました。

### CODE REVIEW

File: modules/cluster-core-kernel/src/ddata/key.rs [hunk 1]
```diff
@@ -64,3 +64,6 @@ pub type PNCounterKey = Key<PNCounter>;
 
 /// Key for [`PNCounterMap`] values.
 pub type PNCounterMapKey<K> = Key<PNCounterMap<K>>;
+
+/// Key for [`VersionVector`] values.
+pub type VersionVectorKey = Key<VersionVector>;
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply modules/cluster-core-kernel/src/ddata/key.rs:69]: 対応しました。`VersionVectorKey` を `version_vector_key.rs` に分離し、`ddata.rs` から re-export する形に変更しました。ローカルでは `cargo test -p fraktor-cluster-core-kernel-rs version_vector --lib` と `./scripts/ci-check.sh ai all` が通過済みです。

File: modules/cluster-core-kernel/src/ddata/version_vector_key.rs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+//! Typed distributed data key for version vector values.
+
+use super::{Key, VersionVector};
+
+/// Key for [`VersionVector`] values.
+pub type VersionVectorKey = Key<VersionVector>;
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply modules/cluster-core-kernel/src/ddata/version_vector_key.rs]: 対応しました。Rust ルールでは pub type alias は 1公開型判定の対象外で、既存の FlagKey / GCounterKey / PNCounterKey / PNCounterMapKey と同じ配置に合わせるのが妥当と判断しました。VersionVectorKey を key.rs の既存 alias 群へ戻し、version_vector_key.rs と個別 re-export を削除しています。ローカルでは cargo test -p fraktor-cluster-core-kernel-rs version_vector --lib、git diff --check、./scripts/ci-check.sh ai all が通っています。

File: modules/cluster-core-kernel/src/ddata/version_vector.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    if removed_version == 0 {
+      return Ok(self.clone());
+    }
+
+    let mut versions = self.versions.clone();
+    versions.remove(removed_node);
+
+    if removed_node != collapse_into {
+      let collapse_version = versions.get(collapse_into).copied().unwrap_or(0);
+      let next = max(collapse_version, removed_version).checked_add(1).ok_or(CounterArithmeticError::Overflow)?;
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply modules/cluster-core-kernel/src/ddata/version_vector.rs]: 指摘は妥当でした。`prune` で removed node の clock を collapse 先へ持ち越さず、collapse 先自身の clock を `+1` するように修正しました。

回帰テストとして removed=100 / collapse=2 のケースを追加し、prune 後の collapse clock が 3 になり、collapse=50 の VersionVector より before のままになることを確認しています。

ローカルでは以下を確認済みです。
- `cargo test -p fraktor-cluster-core-kernel-rs version_vector --lib`
- `git diff --check`
- `./scripts/ci-check.sh ai all`

File: modules/cluster-core-kernel/src/ddata/version_vector.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+      versions.insert(collapse_into.clone(), next);
+    }
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply modules/cluster-core-kernel/src/ddata/version_vector.rs:184]: 確認しましたが、この指摘は現行差分に対しては偽陽性として扱います。

現在の実装は `max(collapse_version, removed_version) + 1` ではなく、removed node を削除したうえで collapse 先自身の clock を `collapse_version + 1` だけ進めています。Pekko参照実装も `ManyVersionVector.prune` では `VersionVector(versions = versions - removedNode) + collapseInto` で、ここでの `+` は `increment` です。そのため removed node の clock を plain max で持ち越す実装にはしていません。

この挙動は直前の回帰テスト `pruning_does_not_reuse_removed_node_version_as_collapse_clock` でも、removed=100 / collapse=2 の prune 結果が collapse=3 になり、collapse=50 より after にならないことを確認しています。

File: modules/cluster-core-kernel/src/ddata/version_vector.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+    let mut versions = self.versions.clone();
+    versions.remove(removed_node);
+
+    if removed_node != collapse_into {
+      let collapse_version = versions.get(collapse_into).copied().unwrap_or(0);
+      let next = collapse_version.checked_add(1).ok_or(CounterArithmeticError::Overflow)?;
+      versions.insert(collapse_into.clone(), next);
+    }
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply modules/cluster-core-kernel/src/ddata/version_vector.rs:184]: 確認しましたが、この指摘は偽陽性として扱います。

Pekko参照実装の `ManyVersionVector.prune` は `VersionVector(versions = versions - removedNode) + collapseInto` で、ここでの `+` は `increment` です。つまり removed node のversionを `max(collapse_version, removed_version)` でcollapse先へ持ち越すのではなく、removed node を削除したうえで collapse先自身のclockを1つ進める挙動です。

`VersionVector` は因果clockであり、`GCounter` の値合算や `PNCounterMap` のdot/tombstoneとは保持すべき意味が異なります。removed側のclockをcollapse先に混ぜると、別nodeの因果履歴をcollapse先nodeの履歴として表現してしまうため、Codex Reviewで指摘された ordering の破壊につながります。

そのため現在の `collapse_version + 1` を維持します。回帰テスト `pruning_does_not_reuse_removed_node_version_as_collapse_clock` はこの意図を固定するためのものです。

File: modules/cluster-core-kernel/src/ddata/version_vector_test.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+fn pruning_does_not_reuse_removed_node_version_as_collapse_clock() {
+  let removed = unique_address(0);
+  let collapse_into = unique_address(1);
+  let vector = vector_from_entries(&[(0, 100), (1, 2)]);
+
+  let pruned = vector.prune(&removed, &collapse_into).expect("collapse version must fit");
+
+  assert_eq!(pruned.version_at(&collapse_into), 3);
+  assert!(pruned.is_before(&vector_from_entries(&[(1, 50)])));
+}
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply modules/cluster-core-kernel/src/ddata/version_vector_test.rs:107]: こちらも同じ理由で偽陽性として扱います。

このテストは、removed node のversionをcollapse先のclockとして再利用しないことを明示する回帰テストです。Pekko参照実装では `versions - removedNode` のあとに `+ collapseInto` しており、`+` はcollapse先nodeのincrementです。そのため `{A:100, B:2}` をAからBへpruneした結果が `{B:3}` になり、`{B:50}` よりbeforeになることは意図した挙動です。

`max(collapse_version, removed_version) + 1` にすると、AのclockをBのclockとして表現してしまい、VersionVectorのnode別因果履歴の意味が崩れます。したがってこのテスト期待値は維持します。

**Comment by j5ik2o** [comment]
レビュー/自動コメント確認済みです。

- CodeRabbit walkthrough / pre-merge checks: 変更概要と事前チェック結果を確認済み。追加対応は不要です。
- Codecov report: patch coverage と project coverage の成功を確認済み。coverage対応は不要です。
- Codex Review: `VersionVector::prune` の因果clock混在指摘は妥当だったため、`2c9d8d815` で修正済みです。
- Devin Review: `max(collapse_version, removed_version) + 1` へ戻す提案はPekko参照実装とVersionVectorの意味に照らして偽陽性と判断し、各threadへ理由を返信してresolve済みです。

現時点では未解決review threadは残していない想定です。
